examples/amm/src/contract.rs

```diff
@@ -9,15 +9,15 @@ use amm::{AmmAbi, AmmError, Message, Operation, Parameters};
 use fungible::{Account, FungibleTokenAbi};
 use linera_sdk::{
     base::{AccountOwner, Amount, ApplicationId, ChainId, WithContractAbi},
-    ensure, Contract, ContractRuntime,
+    ensure, Contract, ContractRuntime, StoreOnDrop,
 };
 use num_bigint::BigUint;
 use num_traits::{cast::FromPrimitive, ToPrimitive};
 
 use self::state::Amm;
 
 pub struct AmmContract {
-    state: Amm,
+    state: StoreOnDrop<Amm>,
```

Comment:
Mmm. Asking users to add this is kinda significant. What if they forget?

Reply:
That is a valid point, and definitely a risk. I think it's mitigated by having the `linera project new` template already including it though.

I do think it's easier to audit if a contract has `StoreOnDrop` than if it's overriding `finalize` :man_shrugging:

Reply:
But most apps are not overriding `finalize`.

Reply:
But all apps need to have `state_mut`, which I think is quite hacky and unclear why it's needed (e.g., will the application's state be changed by the system?). It also forces `Contract` implementations for applications without a state to still have a `state` field. If apps need to have a hack, I'd prefer to make explicit what it does :thinking: